libraries/ESP8266WebServer/src/ESP8266WebServer-impl.h

```diff
@@ -439,44 +432,80 @@ void ESP8266WebServerTemplate<ServerType>::setContentLength(const size_t content
+{
+    return _streamIt(name) && _streamItC(F(": ")) && _streamIt(value) && _streamItC(F("\r\n"));
+}
+
+template <typename ServerType>
+bool ESP8266WebServerTemplate<ServerType>::_sendHeader(int code, const char* content_type, size_t contentLength) {
+    if (   !_streamItC(F("HTTP/1."))
+        || !_streamIt(String(_currentVersion))
+        || !_streamItC(F(" "))
+        || !_streamIt(String(code))
+        || !_streamItC(F(" "))
+        || !_streamIt(responseCodeToString(code))
+        || !_streamItC(F("\r\n")))
+    {
+        return false;
+    }
 
     using namespace mime;
     if (!content_type)
         content_type = mimeTable[html].mimeType;
+    if (!_streamHeader(F("Content-Type"), content_type))
+        return false;
 
-    sendHeader(String(F("Content-Type")), String(FPSTR(content_type)), true);
     if (_contentLength == CONTENT_LENGTH_NOT_SET) {
-        sendHeader(String(FPSTR(Content_Length)), String(contentLength));
+        if (!_streamHeader(Content_Length, String(contentLength)))
+            return false;
     } else if (_contentLength != CONTENT_LENGTH_UNKNOWN) {
-        sendHeader(String(FPSTR(Content_Length)), String(_contentLength));
-    } else if(_contentLength == CONTENT_LENGTH_UNKNOWN && _currentVersion){ //HTTP/1.1 or above client
-      //let's do chunked
-      _chunked = true;
-      sendHeader(String(F("Accept-Ranges")),String(F("none")));
-      sendHeader(String(F("Transfer-Encoding")),String(F("chunked")));
+        if (!_streamHeader(Content_Length, String(_contentLength)))
+            return false;
+    } else if (_contentLength == CONTENT_LENGTH_UNKNOWN && _currentVersion){ //HTTP/1.1 or above client
+        //let's do chunked
+        _chunked = true;
+        if (!_streamHeader(F("Accept-Ranges"), F("none")) || !_streamHeader(F("Transfer-Encoding"), F("chunked")))
+            return false;
     }
+
     if (_corsEnabled) {
-      sendHeader(String(F("Access-Control-Allow-Origin")), String("*"));
+        if (!_streamHeader(F("Access-Control-Allow-Origin"), F("*")))
+            return false;
     }
 
-    if (_keepAlive && _server.hasClient()) { // Disable keep alive if another client is waiting.
-      _keepAlive = false;
+    if (_keepAlive && _server.hasClient()) {
+        // Disable keep alive if another client is waiting.
+        _keepAlive = false;
+    }
+    if (!_streamHeader(F("Connection"), String(_keepAlive ? F("keep-alive") : F("close")))) {
+        return false;
     }
-    sendHeader(String(F("Connection")), String(_keepAlive ? F("keep-alive") : F("close")));
     if (_keepAlive) {
-      sendHeader(String(F("Keep-Alive")), String(F("timeout=")) + HTTP_MAX_CLOSE_WAIT);
+        if (!_streamHeader(F("Keep-Alive"), String(F("timeout=")) + HTTP_MAX_CLOSE_WAIT))
+            return false;
     }
 
+    for (const auto& kv: _userHeaders)
+        if (!_streamHeader(kv.first, kv.second))
+            return false;
+    _userHeaders.clear();
```

Comment:
When will these be cleared if streaming failed?

N.B. I got the impression the current implementation of the webserver does hold some data way longer than needed as opening a web page may sometimes free up quite a bit of memory. (maybe POST data of a previous request???)